stackstate_checks_base/stackstate_checks/base/utils/http_helper.py

```diff
@@ -0,0 +1,1447 @@
+        # Test if the HTTP Response Type is a enum of type HTTPResponseType
+        if self._validate_type is not None and \
+                type(self._validate_type) is HTTPResponseType and \
+                isinstance(self._validate_type, Enum) is True:
+
+            # If the type set is JSON. We then attempt to parse it and see if it is valid
+            if self._validate_type is HTTPResponseType.JSON:
+                try:
+                    data = json.loads(response.content.decode(response.encoding))
+                    if isinstance(data, dict) is False:
+                        raise ValueError()
+
+                except ValueError as e:
+                    errors.append("""Unable to parse the response as JSON.
+                                     The response received was {0}.
+                                     Full error from the JSON parse attempt {1}
+                                     """.format(str(response.content), e))
+
+            elif isinstance(self._validate_type.value, type(response.content)) is False:
+                errors.append("""The response content type does not conform to the validation type. The response
+                                 type is {0}.
+                                 The expected type is {1}
+                                 To fix this you can either modify the type validation or remove it entirely
+                                 """.format(str(type(response.content)), str(self._validate_type.value)))
+
+        # Return all the errors that occurred
+        return errors
+
+
+"""
+    The HTTP Helper Handler is used to create compact methods using most of the function defined in the
+    Connection, Request, Session and Response Helpers
+    Functionality:
+        - Overwrite Connection Helper
+        - Overwrite Request Helper
+        - Overwrite Session Helper
+        - Overwrite Response Helper
+        - GET Request
+        - POST Request
+        - DELETE Request
+        - PATCH Request
+        - PUT Request
+
+    There is 3 ways to approach creating a HTTP Helper
+    - The developer can create the HTTPHelperRequestHandler, HTTPHelperSessionHandler, HTTPHelperResponseHandler and
+      HTTPHelperConnectionHandler.
+      These can then manually be passed down as kwargs to the send function
+    - The developer can create the HTTPHelperRequestHandler, HTTPHelperSessionHandler, HTTPHelperResponseHandler and
+      HTTPHelperConnectionHandler.
+      These can then manually be set within the HTTPHelper class with the overwrite functions
+    - Or it can be left to the internal state where the Helper control and build the state.
+      Thus if the developer use the setters and getter those internal state will be manipulated
+"""
+
+
+class HTTPHelper:
+    _common = HTTPHelperCommon()
+    _connection = HTTPHelperConnectionHandler()
+    _request = HTTPHelperRequestHandler()
+    _session = HTTPHelperSessionHandler()
+    _response = HTTPHelperResponseHandler()
```

Comment:
Why do you create these class variables?

Reply:
This is to allow something like a simple GET or POST function. Those objects needs to be created so that the send function can use them. This makes it easier for the developer to not be forced to manually create them and pass them into the GET function but rather just call the GET function.
There is also functionality to allow the developer to manually overwrite anyone of those if need be.

If these are removed then the developer will always be expected to pass these as parameters, If this is a beter approach then I will gladly remove those constructors.

Reply:
Baked the creation of these class variables into the builder function. The builder function will now create these, apply the correct data and trigger the send. This also reduced mutability